File: product/conference/src/main/java/ee/ctob/service/ParticipantService.java
package ee.ctob.service;

import ee.ctob.access.ConferenceDAO;
import ee.ctob.access.ParticipantDAO;
import ee.ctob.access.RoomDAO;
import ee.ctob.api.Response;
import ee.ctob.api.dto.ParticipantDTO;
import ee.ctob.data.Conference;
import ee.ctob.data.Participant;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import static java.time.LocalDateTime.now;


@Service
public class ParticipantService {
    @Autowired
    ParticipantDAO participantDAO;
    @Autowired
    ConferenceDAO conferenceDAO;
    @Autowired
    RoomDAO roomDAO;

    public ParticipantDTO registration(ParticipantDTO participantDTO) {
        UUID participantUUID = UUID.randomUUID();
        if(conferenceDAO.registerParticipant(participantUUID, participantDTO.getConferenceUUID())==0) {
            return ParticipantDTO.builder()
                    .info("Registration isn't available for this conference")
                    .build();
        }

        Participant participant = participantDAO.saveAndFlush(
                Participant.builder()
                        .created(now())
                        .participantUUID(participantUUID)
                        .validationUUID(UUID.randomUUID())
                        .firstName(participantDTO.getFirstName())
                        .lastName(participantDTO.getLastName())
                        .email(participantDTO.getEmail())
                        .gender(participantDTO.getGender())
                        .dateOfBirth(participantDTO.getDateOfBirth())
                        .build()
        );

        return ParticipantDTO.builder()
                .validationUUID(participant.getValidationUUID())
                .build();
    }

    public ParticipantDTO registrationCancel(ParticipantDTO participantDTO) {
        Participant participant = participantDAO.getParticipantUUID(participantDTO.getValidationUUID());
        if(participant == null) {
            return ParticipantDTO.builder()
                    .info("Participant with this validation doesn't exists")
                    .build();
        }

        UUID participantUUID = participant.getValidationUUID();

        if(conferenceDAO.isAvailableForCancel(participantUUID) == null) {
            return ParticipantDTO.builder()
                    .info("Conference already started or finished")
                    .build();
        }

        if(conferenceDAO.cancelRegistration(participantUUID)==0) {
            return ParticipantDTO.builder()
                    .info("Validation uuid isnt valid")
                    .build();

        }
        return ParticipantDTO.builder()
                .registrationCancel(true)
                .build();
    }

    public ParticipantDTO feedback(ParticipantDTO participantDTO) {
        boolean result = participantDAO.feedback(participantDTO.getValidationUUID(), participantDTO.getFeedback())>0;

        ParticipantDTO.ParticipantDTOBuilder participantDTOBuilder = ParticipantDTO.builder()
                .validationUUID(participantDTO.getValidationUUID())
                .feedbackResult(result);

        if(!result) {
            return participantDTOBuilder.info("Feedback already exists or conference isn't finished").build();
        }

        return participantDTOBuilder.build();
    }

    public ParticipantDTO availableConferences(ParticipantDTO participantDTO) {
        if(now().isAfter(participantDTO.getFrom()) || participantDTO.getUntil().isBefore(participantDTO.getFrom())) {
            return ParticipantDTO.builder()
                    .info("Requested time isn't logical")
                    .build();
        }

        List<Conference> conferenceList = conferenceDAO.findAllAvailableBetween(participantDTO.getFrom(), participantDTO.getUntil());
        if(conferenceList.isEmpty()) {
            return  ParticipantDTO.builder()
                    .info("No conferences is available at this time period")
                    .build();
        }

        List<Response.ConferenceAvailable> conferenceAvailableList = new ArrayList<>();
        for(Conference conference : conferenceList) {
            conferenceAvailableList.add(Response.ConferenceAvailable.builder()
                    .conferenceUUID(conference.getConferenceUUID())
                    .location(roomDAO.getRoomLocationByRoomId(conference.getRoomUUID()))
                    .participantsAmount(conference.getParticipants().size())
                    .info(conference.getInfo())
                    .from(conference.getBookedFrom())
                    .until(conference.getBookedUntil())
                    .build());
        }

        return ParticipantDTO.builder()
                .conferenceAvailableList(conferenceAvailableList)
                .build();
    }
}

File: product/conference/src/main/kotlin/ee/ctob/service/ParticipantService.kt
package ee.ctob.service

import ee.ctob.access.ConferenceDAO
import ee.ctob.access.ParticipantDAO
import ee.ctob.access.RoomDAO
import ee.ctob.api.dto.RequestDTO
import ee.ctob.api.dto.ResponseDTO
import ee.ctob.api.error.BadRequestException
import ee.ctob.api.error.PreconditionsFailedException
import ee.ctob.data.Participant
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime.now
import java.util.*

@Service
@Transactional
open class ParticipantService (
    @Autowired private val participantDAO: ParticipantDAO,
    val conferenceDAO: ConferenceDAO,
    private val roomDAO: RoomDAO
) {

    fun registration(requestDTO: RequestDTO): ResponseDTO {
        val newParticipantUUID = UUID.randomUUID()
        if (conferenceDAO.registerParticipant(newParticipantUUID, requestDTO.conferenceUUID!!) == 0) {
            throw PreconditionsFailedException("Registration isn't available for this conference")
        }

        val participant = participantDAO.saveAndFlush(
            Participant().apply {
                created = now()
                validationUUID = UUID.randomUUID()
                participantUUID = newParticipantUUID
                firstName = requestDTO.firstName
                lastName = requestDTO.lastName
                email = requestDTO.email
                gender = requestDTO.gender
                dateOfBirth = requestDTO.dateOfBirth
            }
        )

        return ResponseDTO(
            validationUUID = participant.validationUUID
        )
    }

    fun registrationCancel(requestDTO: RequestDTO): ResponseDTO {
        val participant = participantDAO.getParticipant(requestDTO.validationUUID!!) ?:
        throw PreconditionsFailedException("Participant with this validation doesn't exists")

        val participantUUID = participant.participantUUID

        conferenceDAO.isAvailableForCancel(participantUUID!!) ?:
        throw PreconditionsFailedException("Conference already started or finished")

        if (conferenceDAO.cancelRegistration(participantUUID) == 0) {
            throw PreconditionsFailedException("Validation uuid isn't valid")
        }

        return ResponseDTO(
            registrationCancel = true
        )
    }

    fun feedback(requestDTO: RequestDTO): ResponseDTO {
        if (participantDAO.feedback(requestDTO.validationUUID!!, requestDTO.feedback!!) == 0) {
            throw PreconditionsFailedException("Feedback already exists or conference isn't finished")
        }

        return ResponseDTO(
            validationUUID = requestDTO.validationUUID,
            feedbackResult = true
        )
    }

    fun availableConferences(requestDTO: RequestDTO): ResponseDTO {
        if (now().isAfter(requestDTO.from) || requestDTO.until!!.isBefore(requestDTO.from)) {
            throw BadRequestException(400, "Requested time isn't logical")
        }

        val conferenceList = conferenceDAO.findAllAvailableBetween(requestDTO.from!!, requestDTO.until!!)

        if(conferenceList.isEmpty()) {
            throw PreconditionsFailedException("No conferences are available at this time period")
        }

        val conferenceAvailableList = conferenceList.map { conference ->
            ResponseDTO.ConferenceAvailableDTO(
                conferenceUUID = conference.conferenceUUID,
                location = roomDAO.getRoomLocationByRoomId(conference.roomUUID!!),
                participantsAmount = conference.participants?.size,
                info = conference.info,
                from = conference.bookedFrom,
                until = conference.bookedUntil
            )
        }

        return ResponseDTO(
            conferenceAvailableList = conferenceAvailableList
        )
    }
}
